api/renter_test.go

```diff
@@ -1899,3 +1899,212 @@ func TestContractorHostRemoval(t *testing.T) {
 		t.Log("downloaded file and uploaded file do not match")
 	}
 }
+
+// TestExhaustedContracts verifies that the contractor renews contracts which
+// run out of funds before the period elapses.
+func TestExhaustedContracts(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+	t.Parallel()
+	st, err := createServerTester(t.Name())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer st.server.panicClose()
+
+	// set a very high price for the host so we use up the entire funds of the
+	// contract
+	err = st.acceptContracts()
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = st.setHostStorage()
+	if err != nil {
+		t.Fatal(err)
+	}
+	settings := st.host.InternalSettings()
+	settings.MinUploadBandwidthPrice = types.SiacoinPrecision.Div64(10)
+	err = st.host.SetInternalSettings(settings)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = st.announceHost()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Set an allowance for the renter, allowing a contract to be formed.
+	allowanceValues := url.Values{}
+	testPeriod := "1000"
+	allowanceValues.Set("funds", types.SiacoinPrecision.Mul64(1000).String())
+	allowanceValues.Set("period", testPeriod)
+	err = st.stdPostAPI("/renter", allowanceValues)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// wait until we have a contract
+	err = retry(500, time.Millisecond*50, func() error {
```

Comment:
phasing this out for `build.Retry`

Reply:
it's the same function just accessible to everyone instead of silo'd to the api package

Reply:
Ah yeah, glad we have that merged